compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1302,6 +1302,14 @@ impl<'a> Parser<'a> {
         maybe_recover_from_interpolated_ty_qpath!(self, true);
         maybe_whole_expr!(self);
 
+        // Recover from `pub let` inside a function
+        if self.prev_token.is_keyword(kw::Pub) && self.may_recover() {
```

Comment:
This doesn't check that the current token is a `let` token. 

Also, can we move this to something like `parse_stmt`?

Reply:
`parse_stmt` doesn't call in this case

Reply:
I think this can be added to `parse_stmt_without_recovery`, which calls `parse_expr`. You just need to add a new `else if`.